draft-ietf-quic-transport.md

```diff
@@ -2721,6 +2721,66 @@ between endpoints.  Application protocols SHOULD define rules for handling
 streams that are prematurely cancelled by either endpoint.
 
 
+# Alternative Initial Set {#alternative-initial}
+
+In order to avoid ossification of the cleartext and obfuscated fields of QUIC
+packets, a server can announce an alternative set of initial values to be used,
+which is comprised of:
+
+* Alternative version number; a 32-bit unsigned number that is to be presented
+  on wire in place of the version number specified in this document.  This value
+  MUST NOT be a reserved version ({{versions}}).
+
+* Packet type modifier; a two-bit value that is to be applied as a bit-wise
+  exclusive or (XOR) to the most significant bits of the Initial, Handshake,
+  0-RTT, and Retry packets. This XOR is applied after the packets are encrypted
+  and before they are decrypted.
+
+* Alternative initial salt; a 16-byte binary blob that is to be used in place of
+  the initial salt defined in section 5.2 of {{QUIC-TLS}}.
+
+A server advertises these values using a NEW_TOKEN frame {{frame-new-token}}.
+The token MUST include or associated with the alternative version number with
+which it can be used.
+
+Typically, a server would pre-allocate a set of unused version numbers as the
+alternative version numbers, associating each of those version numbers with a
+packet type modifier chosen at random.  Then, when issuing a token using a
+NEW_TOKEN frame, the server generates the alternative initial salt by calling a
+pseudo-random function, embeds that initial salt into the token which is then
+encrypted, and sends a NEW_TOKEN frame that comprises of the generated token and
+the alternative initial set.
+
+When the client reconnects to the server by using the provided token and the
+alternative initial set, the server first checks if the version number field of
+the incoming packet contains one of the alternative version numbers it
+advertises, then if that is the case, applies the corresponding packet type
+modifier to recover the correct packet type.  If the recovered packet type is an
+Initial packet and that packet contains a NEW_TOKEN token, the server decrypts
+the embedded token and recovers the alternative initial salt, uses that to
+decrypt the payload of the Initial packet.
+
+When the server is incapable of determining the alternative initial salt, it can
+send a Version Negotiation packet that instructs the client to use the default
```

Comment:
In the current spec, the client fails the connection upon receipt of a Version Negotiation packet. So this has a dependency on the version downgrade prevention extension.

But assuming the presence of a version negotiation mechanism with downgrade prevention, I think we have an issue here. If the server rotates the key it uses to encrypt the alternative initial salt, and then receives a token it can no longer decrypt, it sends a version negotiation packet, and the client tries again with QUICv1. This is likely to trip the version downgrade prevention mechanism - how do we distinguish between a server that has forgotten their key and an attacker that is trying to force a downgrade to QUICv1?

Reply:
This PR has a built-in downgrade prevention mechanism.

The requirement we have in this PR is that each NEW_TOKEN token that the server issues is linked to a particular alternative version number.

When a server receives an Initial packet with an alternative version number it recognizes but with a token it cannot use, the server sends a VN. The client reconnects using V1, but the text in the PR requires the client to continue using the token. Then, when the server receives the V1 packet with the token, and can in fact use the token and finds out that the token was in fact being issued for a version other than V1, the server generates a VERSION_NEGOTIATION_ERROR.

In short, it is a downgrade attack if VN was issued while the token was usable.

Reply:
Let's assume the presence of an on-path attacker that can drop, forge and rewrite packets. This attacker is interested in parsing the SNI and ALPN. The attacker lets the client's first QUICv1 connection through, and the client gets a token with alternate version and salts, under the cover of encryption. Some time later, the client wants to open a second connection and sends an obfuscated initial with the new version and salt. The attacker can pretty easily figure out that the first packet is an Initial even with the 2bit XOR. The attacker modifies some bits in the token, and recomputes the checksum then sends the modified packet to the server, leaving the version intact. The server verifies the checksum then fails to decrypt the token. The server sends VN. The client then retries with QUICv1 using its original token. The attacker performs the same modification on that token, and recomputes the checksum and initial auth tag, then sends the packet to the server. The server will again fail to decrypt the token, it sends a retry, the client tries again with the retry token, which the attacker leaves alone, and the connection successfully establishes with QUICv1. The **attacker has successfully forced a downgrade** to QUICv1 and was able to extract SNI and ALPN.

Reply:
Yes, this offers an almost trivial downgrade attack, but this isn't a scenario that we agreed to cover.  It does seem to be a little bit *too* trivial an attack though.

Reply:
@DavidSchinazi That's obviously the case. Thank you for pointing that out.

I think what the PR fails to state that a Retry token needs to inherit the properties of the NEW_TOKEN token included in the Initial packet that trigerred the Retry. One way doing that is to embed the NEW_TOKEN token in the Retry token.

The reason why we need the NEW_TOKEN token to be embedded in the Retry token is not only because there is an attack that you describe. It is necessary because when the server receives the first Initial packet from client that carries the alternative version, and responds with a Retry packet that carries that alternative version, and the client in response sends an Initial packet that carries the Retry token using an Initial packet with the alternative version, the server needs to extract the alternative initial salt (and other information) from the Retry token.

I'd update the PR to clarify that.

Reply:
@martinthomson I think the design principle behind this PR that I have are:
* passive attack does not reveal the payload
* active attack can lead to the payload being revealed, but is detectable by the endpoints and thereby disrupts the handshake
* handshake falls back to v1 when the server forgets the token encryption key

As I think the issue pointed out by @DavidSchinazi is addressable (see my comment above), I think we might stick to the principle and how it evolves. If we conclude that the design based on the principle is simple enough, then we'd be fine. If not, we can consider the alternatives.

Reply:
373257b clarifies that the alternative initial set would (or should) be available through the Retry token. The text is kept informative, as the PR no longer allows a downgrade.